### PR TITLE
Issue #991 Cancel auto transition when trasition happend in Open state

### PR DESCRIPTION
**Pull request by dlsrb6342**

### CODE REVIEW

**Comment by RobWin** [comment]
Thanks,
could you please add a test to verify the expected behavior?

**Comment by dlsrb6342** [comment]
@RobWin 
Added tests. Please have a look :)

**Comment by RobWin** [comment]
Is it possible to use the MockClock instead of Thread.sleep? Not sure if the scheduler supports it.

**Comment by dlsrb6342** [comment]
@RobWin 

```java
@Test
public void shouldCancelAutoTransition() {
    ScheduledFuture<?> mockFuture = mock(ScheduledFuture.class);
    doReturn(mockFuture)
        .when(schedulerMock).schedule(any(Runnable.class), any(Long.class), any(TimeUnit.class));

    // Auto transition scheduled
    circuitBreaker.transitionToOpenState();

    // Auto transition should be canceled
    circuitBreaker.transitionToForcedOpenState();

    // Verify scheduled future is canceled
    then(mockFuture).should(times(1)).cancel(true);
}

```

mocked future and verified `future.cancel()` is called only one time. 
I think this is enough to test behavior. What do you think?

**Comment by RobWin** [comment]
But with a MockClock wr can also mock the time and dont need thread sleep. Please have a look at the other CB state transition tests.

**Comment by dlsrb6342** [comment]
I think `ScheduledExecutorService` doesn't support `MockClock`. 

Other auto transition tests are also just mocking `ScheduledExecutorService`.
https://github.com/resilience4j/resilience4j/blob/master/resilience4j-circuitbreaker/src/test/java/io/github/resilience4j/circuitbreaker/internal/CircuitBreakerAutoTransitionStateMachineTest.java

File: resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/internal/CircuitBreakerStateMachine.java [hunk 1]
```diff
@@ -293,6 +291,9 @@ private void stateTransition(State newState,
         UnaryOperator<CircuitBreakerState> newStateGenerator) {
         CircuitBreakerState previousState = stateReference.getAndUpdate(currentState -> {
             StateTransition.transitionBetween(getName(), currentState.getState(), newState);
+            if (currentState instanceof OpenState) {
```
Comment (RobWin) [inline resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/internal/CircuitBreakerStateMachine.java:294]: I'm not happy that this state-specific code is in generic method of the CircuitBreakerStateMachine.
Maybe we can add a `postTransitionHook` or `preTransitionHook` to states which are invoked here in a generic form?

Comment (dlsrb6342) [reply resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/internal/CircuitBreakerStateMachine.java:294]: Yes right. that would be better. I'll add `preTransitionHook`.

File: resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/internal/CircuitBreakerStateMachine.java [hunk 2]
```diff
@@ -695,12 +706,23 @@ public CircuitBreakerMetrics getMetrics() {
...
+
         private void toHalfOpenState() {
             if (isOpen.compareAndSet(true, false)) {
                 transitionToHalfOpenState();
             }
         }
 
+        private void cancelAutomaticTransitionToHalfOpen() {
+            if (transitionToHalfOpenFuture != null) {
+                transitionToHalfOpenFuture.cancel(true);
```
Comment (RobWin) [inline resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/internal/CircuitBreakerStateMachine.java:722]: what happens if the future is already completed?

Comment (RobWin) [reply resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/internal/CircuitBreakerStateMachine.java:722]: does it really work? 
https://stackoverflow.com/questions/29181905/future-cancel-is-not-cancelling-the-scheduled-execution-of-scheduledexecutorse

Comment (dlsrb6342) [reply resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/internal/CircuitBreakerStateMachine.java:722]: Yes. Needed to check whether it is done or is cancelled already.
Thank you.

**Comment by RobWin** [comment]
Thanks

**Pull request merged**